src/main/java/io/github/thebusybiscuit/slimefun4/core/networks/energy/EnergyNet.java

```diff
@@ -139,138 +149,141 @@ public void tick(Block b) {
+            int capacity = component.getCapacity();
 
-            if (available > 0) {
-                if (available > capacity) {
-                    ChargableBlock.setUnsafeCharge(generator, capacity, false);
-                    available -= capacity;
+            if (remainingEnergy > 0) {
+                if (remainingEnergy > capacity) {
+                    component.setCharge(l, capacity);
+                    remainingEnergy -= capacity;
                 }
                 else {
-                    ChargableBlock.setUnsafeCharge(generator, available, false);
-                    available = 0;
+                    component.setCharge(l, remainingEnergy);
+                    remainingEnergy = 0;
                 }
             }
             else {
-                ChargableBlock.setUnsafeCharge(generator, 0, false);
+                component.setCharge(l, 0);
             }
         }
     }
 
-    private int tickAllGenerators(Map<Location, Integer> generatorsWithCapacity, LongConsumer timeCallback) {
+    private int tickAllGenerators(LongConsumer timings) {
         Set<Location> exploded = new HashSet<>();
         int supply = 0;
 
-        for (Location source : generators) {
+        for (Map.Entry<Location, EnergyNetComponent> entry : generators.entrySet()) {
             long timestamp = SlimefunPlugin.getProfiler().newEntry();
-            Config config = BlockStorage.getLocationInfo(source);
-            SlimefunItem item = SlimefunItem.getByID(config.getString("id"));
+            Location l = entry.getKey();
+            EnergyNetComponent component = entry.getValue();
 
-            if (item instanceof EnergyNetProvider) {
+            if (component instanceof EnergyNetProvider) {
+                SlimefunItem item = (SlimefunItem) component;
                 try {
-                    EnergyNetProvider provider = (EnergyNetProvider) item;
-                    int energy = provider.getGeneratedOutput(source, config);
+                    EnergyNetProvider provider = (EnergyNetProvider) component;
+                    Config config = BlockStorage.getLocationInfo(l);
+                    int energy = provider.getGeneratedOutput(l, config);
 
-                    if (provider.getCapacity() > 0) {
-                        generatorsWithCapacity.put(source, provider.getCapacity());
-                        String charge = config.getString("energy-charge");
-
-                        if (charge != null) {
-                            energy += Integer.parseInt(charge);
-                        }
+                    if (provider.isChargeable()) {
+                        energy += provider.getCharge(l);
                     }
 
-                    if (provider.willExplode(source, config)) {
-                        exploded.add(source);
-                        BlockStorage.clearBlockInfo(source);
+                    if (provider.willExplode(l, config)) {
+                        exploded.add(l);
```

Comment:
Make a variable name less verbose, that's not like you Cookie

Reply:
Do you prefer loc over l...